### DISCUSSION TITLE
[@types/puppeteer] Incorrect return type for $eval and $$eval

### DISCUSSION THREAD

**Discussion by FeldrinH** [DefinitelyTyped/DefinitelyTyped#60673] [Issues with a @types package] [answered]
The types for Puppeteer incorrectly claim that $eval and $$eval will wrap elements returned from the parameter function in ElementHandles. In reality, they only return serializable objects, so returning an Element results in {} (an empty object) being returned. See https://github.com/puppeteer/puppeteer/issues/2609 and [the Puppeteer docs themselves](https://github.com/puppeteer/puppeteer/blob/master/docs/api.md#pageevalselector-pagefunction-args-1) for more details.

This issue was previously filed under #44403 and seems to still be unresolved. More details can be found on that page.

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment, marked as answer]
Which version? There are many versions around and Puppeteer was migrated to native TS already

**Reply by FeldrinH** [reply]
In my particular case I was using the deno port of puppeteer (https://deno.land/x/puppeteer@14.1.1). However, it appears that the bug was introduced somewhere in the deno port itselt, not in the upstream puppeteer or DefinitelyTyped libraries, so I will close this for now.

**Reply by peterblazejewicz** [reply]
thanks for the info!
